src/NUnitEngine/nunit.engine/Services/TestAgency.cs

```diff
@@ -31,57 +32,131 @@ public partial class TestAgency : ITestAgency, IService
 
+        private readonly List<IAgentLauncher> _launchers = new List<IAgentLauncher>();
+
         // Transports used for various target runtimes
         private TestAgencyRemotingTransport _remotingTransport; // .NET Framework
         private TestAgencyTcpTransport _tcpTransport; // .NET Standard 2.0
 
         internal virtual string RemotingUrl => _remotingTransport.ServerUrl;
         internal virtual string TcpEndPoint => _tcpTransport.ServerUrl;
 
-        public TestAgency()
+        public TestAgency() : this("TestAgency", 0)
+        {
+        }
+
+        public TestAgency(string uri, int port)
         {
-            var uri = "TestAgency-" + Guid.NewGuid();
-            var port = 0;
             _remotingTransport = new TestAgencyRemotingTransport(this, uri, port);
             _tcpTransport = new TestAgencyTcpTransport(this, port);
         }
 
-        public void Register(ITestAgent agent)
+        #region ITestAgentInfo Implementation
+
+        /// <summary>
+        /// Gets a list containing <see cref="TestAgentInfo"/> for all available agents.
+        /// </summary>
+        public IList<TestAgentInfo> GetAvailableAgents()
         {
-            _agentStore.Register(agent);
+            var agents = new List<TestAgentInfo>();
+
+            foreach (var launcher in _launchers)
+                agents.Add(launcher.AgentInfo);
+
+            return agents;
         }
 
-        public ITestAgent? GetAgent(TestPackage package)
+        #endregion
+
+        #region ITestAgentProvider Implementation
+
+        /// <summary>
+        /// Gets a list containing <see cref="TestAgentInfo"/> for any available agents,
+        /// which are able to handle the specified package.
+        /// </summary>
+        /// <param name="package">A Testpackage</param>
+        /// <returns>
+        /// A list of suitable agents for running the package or an empty
+        /// list if no agent is available for the package.
+        /// </returns>
+        public IList<TestAgentInfo> GetAgentsForPackage(TestPackage targetPackage)
         {
-            if (_runtimeService is null || _availableRuntimeService is null)
-                throw new InvalidOperationException("TestAgency needs to be Started first");
+            Guard.ArgumentNotNull(targetPackage, nameof(targetPackage));
+
+            // Initialize lists with ALL available agents
+            var availableAgents = new List<TestAgentInfo>(GetAvailableAgents());
+            var validAgentNames = new List<string>(availableAgents.Select(info => info.AgentName));
 
+            // Look at each included assembly package to see if any names should be removed
+            foreach (var assemblyPackage in targetPackage.Select(p => p.IsAssemblyPackage()))
+            {
+                // Collect names of agents that work for each assembly
+                var agentsForAssembly = new List<string>();
+                foreach (var launcher in _launchers)
```

Comment:
No point testing launchers which are already excluded:
```suggestion
                foreach (var launcher in _launchers.Where(l => validAgentNames.Contains(l.AgentInfo.AgentName)))
```

Reply:
But `validAgentNames` contains all the names and doesn't exclude anything. Perhaps the naming is wrong.